File: db4o.net/Db4oBench/Bench/Logging/Replay/LogReplayer.cs
/* Copyright (C) 2004 - 2007 Versant Inc. http://www.db4o.com */
using System;
using System.IO;
using System.Collections;
using Sharpen.Util;
using Db4objects.Db4o.IO;
using Db4objects.Db4o.Foundation;

using Db4objects.Db4o.Bench.Logging.Replay.Commands;

namespace Db4objects.Db4o.Bench.Logging.Replay
{
    public class LogReplayer
    {
        private readonly String _logFilePath;
        private readonly IBin _bin;
        private readonly ISet _commands;
        private readonly Hashtable _counts;

        public LogReplayer(String logFilePath, IBin bin, ISet commands)
        {
            _logFilePath = logFilePath;
            _bin = bin;
            _commands = commands;
            _counts = new Hashtable();
            foreach (object com in commands)
            {
                _counts[com] = (Int64)0;
            }
        }

        public LogReplayer(String logFilePath, IBin bin): this(logFilePath, bin, LogConstants.AllEntries())
        {
        }

        public void ReplayLog()
        {
            PlayCommandList(ReadCommandList());
        }

        public List4 ReadCommandList()
        {
            List4 list = null;
            StreamReader reader = new StreamReader(_logFilePath);
            String line = null;
            while ((line = reader.ReadLine()) != null)
            {
                IIoCommand ioCommand = ReadLine(line);
                if (ioCommand != null)
                {
                    list = new List4(list, ioCommand);
                }
            }
            reader.Close();
            return list;
        }

        public void PlayCommandList(List4 commandList)
        {
            while (commandList != null)
            {
                IIoCommand ioCommand = (IIoCommand)commandList._element;
                ioCommand.Replay(_bin);
                commandList = commandList._next;
            }
        }


        private IIoCommand ReadLine(String line)
        {
            String commandName;
            if ((commandName = AcceptedCommandName(line)) != null)
            {
                IncrementCount(commandName);
                return CommandForLine(line);
            }
            return null;
        }

        private String AcceptedCommandName(String line)
        {
            if (line.Length == 0)
            {
                return null;
            }
            foreach (String commandName in _commands)
            {
                if (line.StartsWith(commandName))
                {
                    return commandName;
                }
            }
            return null;
        }

        private IIoCommand CommandForLine(String line)
        {
			if (line.StartsWith(LogConstants.ReadEntry))
            {
				var param = Parameter(LogConstants.ReadEntry, line);
				return new ReadCommand(param.pos, param.len);
            }

            if (line.StartsWith(LogConstants.WriteEntry))
            {
                var param = Parameter(LogConstants.WriteEntry, line);
                return new WriteCommand(param.pos, param.len);
            }
            
			if (line.StartsWith(LogConstants.SyncEntry))
            {
                return new SyncCommand();
            }

            return null;
        }


        private Param Parameter(String command, String line)
        {
            return Parameter(command.Length, line);
        }

		private Param Parameter(int start, String line)
		{
			String[] paramStr = line.Substring(start).Split(' ');
			long pos = long.Parse(paramStr[0]);
			int len = int.Parse(paramStr[1]);
			return new Param(pos, len);
		}

        private void IncrementCount(String key)
        {
            long count = (Int64)_counts[key];
            _counts[key] = (Int64)(count + 1);
        }

        public Hashtable OperationCounts()
        {
            return _counts;
        }
    }

	internal class Param 
	{
		public long pos;
		public int len;

		public Param(long pos, int len) {
			this.pos = pos;
			this.len = len;
		}
	}
}

File: db4otools/src/com/db4o/bench/logging/replay/LogReplayer.java
/* Copyright (C) 2004 - 2007 Versant Inc. http://www.db4o.com */

package com.db4o.bench.logging.replay;

import java.io.*;
import java.util.*;

import com.db4o.bench.logging.*;
import com.db4o.bench.logging.replay.commands.*;
import com.db4o.foundation.*;
import com.db4o.io.*;
/**
 * @sharpen.ignore
 */
public class LogReplayer {
	
	private String _logFilePath;
	private Bin _bin;
	private Set _commands;
	private Map _counts;
	
	public LogReplayer(String logFilePath, Bin bin, Set commands) {
		_logFilePath = logFilePath;
		_bin = bin;
		_commands = commands;
		_counts = new HashMap();
		Iterator it = commands.iterator();
		while (it.hasNext()) {
			_counts.put(it.next(), new Long(0));
		}
	}
	
	public LogReplayer(String logFilePath, Bin bin) {
		this(logFilePath, bin, LogConstants.allEntries());
	}
	
	
	public void replayLog() throws IOException {
		playCommandList(readCommandList());
	}
	
	public List4 readCommandList() throws IOException {
		List4 list = null;
		BufferedReader reader = new BufferedReader(new FileReader(_logFilePath));
		String line = null;
		while ( (line = reader.readLine()) != null ) {
			IoCommand ioCommand = readLine(line);
			if(ioCommand != null){
				list = new List4(list, ioCommand);
			}
		}
		reader.close();
		return list;
	}
	
	public void playCommandList(List4 commandList){
		while(commandList != null){
			IoCommand ioCommand = (IoCommand) commandList._element;
			ioCommand.replay(_bin);
			commandList = commandList._next;
		}
	}
	
	
	private IoCommand readLine(String line) {
		String commandName;
		if ((commandName = acceptedCommandName(line)) != null) {
			incrementCount(commandName);
			return commandForLine(line);
		}
		return null;
	}

	private String acceptedCommandName(String line) {
		if (line.length() == 0) {
			return null;
		}
		Iterator it = _commands.iterator();
		while (it.hasNext()) {
			String commandName = (String)it.next();
			if ( line.startsWith(commandName) ) {
				return commandName;
			}
		}
		return null;
	}
	
	private IoCommand commandForLine(String line) {
		if (line.startsWith(LogConstants.READ_ENTRY)) {
			Param param = parameter(LogConstants.READ_ENTRY,  line);
			return new ReadCommand(param.pos, param.len);
		}
		if ( line.startsWith(LogConstants.WRITE_ENTRY) ) {
			Param param = parameter(LogConstants.WRITE_ENTRY,  line);
			return new WriteCommand(param.pos, param.len);
		}
		if ( line.startsWith(LogConstants.SYNC_ENTRY) ) {
			return new SyncCommand();
		}
		
		return null;
	}

	
	private Param parameter(String command, String line){
		return parameter(command.length(),  line);
	}

	private Param parameter(int start, String line) {
		String[] paramStr = line.substring(start).split(" ");
		long pos = Long.parseLong(paramStr[0]);
		int len = Integer.parseInt(paramStr[1]);
		return new Param(pos, len);
	}

	private static class Param {
		public final long pos;
		public final int len;

		public Param(long pos, int len) {
			this.pos = pos;
			this.len = len;
		}
	}
	
	private void incrementCount(String key) {
		long count = ((Long)_counts.get(key)).longValue();
		_counts.put(key, new Long(count+1));
	}
	
	public Map operationCounts() {
		return _counts;
	}
}
